---
title: "Ortholog genes with nsTE"
author: "Steve Pederson & Lu Zeng"
date: "Feb 2018"
output: html_document
---

```{r, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

# Load packages & Data

## Packages
If you don't have the packages `funsForLu` or `spShortcuts` installed, these can be installed using `library(devtools)` then `install_github("steveped/funsForLu")` and `install_github("steveped/spShortcuts")`

# Setup

```{r loadPackages, message=FALSE, warning=FALSE}
library(ggplot2)
library(stringr)
library(reshape2)
library(matrixStats)
library(VennDiagram)
library(knitr)
library(xtable)
library(scales)
library(boot)
library(lmtest)
library(dplyr)
library(magrittr)
library(parallel)
library(funsForLu)
library(multcomp)
library(pander)
nThreads <- min(16, detectCores() - 1)
```

Load the data
```{r loadData}
tpmFile <- file.path(".", "hg_3nonRTEortholog.txt")
tpmData <- read.delim(tpmFile, sep="\t", stringsAsFactors = FALSE) %>% 
	         tbl_df
```

Collect the elements by type and region
```{r elements}
elements <- grep("IDs", colnames(tpmData), value = TRUE) %>% 
	str_split_fixed(pattern="_", 2) %>%
	as.data.frame(stringsAsFactors = FALSE) %>%
	dplyr::select(type = V1) %>%
	lapply(unique) %>%
	lapply(sort)
nElements <- length(elements$type)
```

Collect the tissues
```{r tissues}
tissues <- c("br", "ht", "kd", "lv", "ts")
nTissues <- length(tissues)
samples <- grep("^hgs", colnames(tpmData), value = TRUE)
```

Add columns of logical values to the main data object which denote the element type and region
```{r addElements}
tpmData %<>%
	bind_cols(sapply(unlist(elements) %>% 
									 	as.vector, 
									 function(x){dplyr::select(tpmData, contains(x)) %>% 
									 		rowSums %>% 
									 		as.logical},
									 simplify = FALSE)) %>%
	mutate(nElTypes = nonRTELINEHg + nonRTESINEHg + nonRTEERVHg)
```

Add further columns which convert the log(TPM) values to binary ```TRUE``` or ```FALSE``` variables for gene expression.
Genes for which log(TPM)$=-\infty$ have TPM$=0$ and will be set to ```FALSE```

```{r addLogical, results='hide'}
grep("^hgs", colnames(tpmData), value = TRUE) %>%
	lapply(function(x){
		tpmData[[paste0(x, "_TF")]] <<- as.logical(exp(tpmData[[x]]))
	})
```

Setup some plotting parameters:

- Use a `labeller()` function for faceting
- Set the default theme to `theme_bw()`

```{r plotParams}
tissueLabels <- as_labeller(c(br = "Brain",
												 ht = "Heart",
												 kd = "Kidney",
												 lv = "Liver",
												 ts = "Testes",
												 Expression = "Expression",
												 Detection = "p(Detection)"))
theme_set(theme_bw() +
						theme(axis.text.x = element_text(angle = 90),
									plot.title = element_text(hjust = 0.5)))
```


# Setup For The Bootstrap Procedure

During the weighted bootstrap procedure, genes will be assigned to groups based on their TE structure.
In order to address any effects on TPM counts due to gene length, the TE groups will be further divided based on length quantiles.
The mean difference in length will also be recorded to ensure that the length distributions are matched.


```{r makeBins}
nBins <- 10
tpmData %<>%
	mutate(lengthBin = findInterval(length, 
																	quantile(length, 
																					 probs = seq(0, 1, length.out = nBins + 1)), 
																	rightmost.closed=TRUE),
				 logLength = log(length))
```

Confidence intervals will be formed for the difference in the probability of a gene being expressed, and for the difference in expression levels due to the presence of each TE within each Genomic Region.
However, in order to control the Type 1 Error at the level $\alpha = 0.05$, each set of $m$ intervals needs to be adjusted, with Bonferroni's method being the simplest approach.
Thus a set of $1-\alpha / m$ Confidence intervals will be formed for each genomic region.

```{r testVals}
m <- nTissues*nElements
alpha <- 0.05/m
nBoot <- 5000L
testVals <- c("logLength", samples, paste0(samples,"_TF"))
```

- In order to provide the number of samples to accurately construct these confidence intervals, the analysis was performed using `r nBoot` bootstrap iterations.
- We can then merge these across replicates within each tissue.
- We also need to check the consistency between replicates within each tissue

# The bootstrap procedure

As each combination of elements & regions may require a different number of genes to be samples, create an empty object for recording these numbers.

```{r}
nGenes <- c()
```


### LINE Elements

Inspect the elements which are LINE Repeats

```{r LINEHgData}
LINEHgData <- tpmData %>%
	mutate(
		elBins = dplyr::select(tpmData, 
													 one_of(grep("nonRTELINEHg", elements$type, invert = TRUE, value = TRUE))) %>%
			binsFromCols, 
		elBins = paste(elBins, lengthBin, sep = "_"))
```

Check we have enough genes to bootstrap:

```{r echo=FALSE}
LINEHgData %>% group_by(nonRTELINEHg) %>% tally() %>% pander()
```

```{r}
nGenes$LINEHg <- 1000L
```

Set the number of bootstrapped genes to `nGenes = ``r nGenes$LINEHg`


```{r bootLINEHg, cache = TRUE}
bootLINEHgOrtho <- mclapply(testVals, 
													 bootMeans, 
													 data = LINEHgData, 
													 testIds = filter(LINEHgData, nonRTELINEHg)$ID, 
													 refIds = filter(LINEHgData, !nonRTELINEHg)$ID,
													 binCol = "elBins",
													 nBoot = nBoot,
													 nGenes = nGenes$LINEHg,
													 maxP = 0.01,
													 mc.cores = nThreads)
names(bootLINEHgOrtho) <- testVals
save(bootLINEHgOrtho, file = "bootnonRTELINEHgOrtho.RData")
```

Check the similarity between replicates for LINEs

```{r, fig.cap = "**LINE Elements**. Visualisation of the differences between replicates within each tissue type. The top row indicates differences between samples for detection of expression, whilst the bottom row indicates differences in expression levels."}
bootLINEHgOrtho %>%
	lapply(extract2, "samples") %>%
	as.data.frame() %>%
	melt() %>%
	as_data_frame() %>%
	mutate(type = gsub("hgs.(br|ht|lv|kd|ts).+", "\\1", variable),
				 TF = c("Expression", "Detection")[grepl("TF", variable) + 1],
				 sample = gsub("_TF", "", variable)) %>%
	filter(type != "logLength") %>%
	ggplot(aes(sample, value, fill = type)) +
	geom_hline(yintercept = 0) +
	geom_boxplot() +
	facet_grid(TF ~ type, scales = "free", space = "free_x", labeller = tissueLabels) +
	labs(title = "Ortholog: LINE Elements",
			 x = "Sample",
			 y = "Change") +
	guides(fill = FALSE) 
```


### SINE Elements

Inspect the elements which are SINEs

```{r SINEHgData}
SINEHgData <- tpmData %>%
	mutate(
		elBins = dplyr::select(tpmData, 
													 one_of(grep("nonRTESINEHg", elements$type, invert = TRUE, value = TRUE))) %>%
			binsFromCols, 
		elBins = paste(elBins, lengthBin, sep = "_")) 
```

Check we have enough genes to bootstrap:

```{r echo=FALSE}
SINEHgData %>% group_by(nonRTESINEHg) %>% tally() %>% pander()
```

```{r}
nGenes$SINEHg <- 1000L
```

Set the number of bootstrapped genes to `nGenes = ``r nGenes$SINEHg`


```{r bootSINEHg, cache = TRUE}
bootSINEHgOrtho <- mclapply(testVals, 
													 bootMeans, 
													 data = SINEHgData, 
													 testIds = filter(SINEHgData, nonRTESINEHg)$ID, 
													 refIds = filter(SINEHgData, !nonRTESINEHg)$ID,
													 binCol = "elBins",
													 nBoot = nBoot,
													 nGenes = nGenes$SINEHg,
													 maxP = 0.01,
													 mc.cores = nThreads)
names(bootSINEHgOrtho) <- testVals
save(bootSINEHgOrtho, file = "bootnonRTESINEHgOrtho.RData")
```

Check the similarity between replicates for SINEs

```{r, fig.cap = "**SINE Elements**. Visualisation of the differences between replicates within each tissue type. The top row indicates differences between samples for detection of expression, whilst the bottom row indicates differences in expression levels."}
bootSINEHgOrtho %>%
	lapply(extract2, "samples") %>%
	as.data.frame() %>%
	melt() %>%
	as_data_frame() %>%
	mutate(type = gsub("hgs.(br|ht|lv|kd|ts).+", "\\1", variable),
				 TF = c("Expression", "Detection")[grepl("TF", variable) + 1],
				 sample = gsub("_TF", "", variable)) %>%
	filter(type != "logLength") %>%
	ggplot(aes(sample, value, fill = type)) +
	geom_hline(yintercept = 0) +
	geom_boxplot() +
	facet_grid(TF ~ type, scales = "free", space = "free_x", labeller = tissueLabels) +
	labs(title = "Ortholog: SINE Elements",
			 x = "Sample",
			 y = "Change") +
	guides(fill = FALSE) 
```


### ERV Elements

```{r ERVHgData}
ERVHgData <- tpmData %>%
	mutate(
		elBins = dplyr::select(
			tpmData, 
			one_of(grep("nonRTEERVHg", elements$type, invert = TRUE, value = TRUE))
		) %>%
			binsFromCols, 
		elBins = paste(elBins, lengthBin, sep = "_")) 
```


Check we have enough genes to bootstrap:

```{r, echo=FALSE}
ERVHgData %>% group_by(nonRTEERVHg) %>% tally() %>% pander()
```


```{r}
nGenes$ERVHg <- 1000L
```


Set the number of bootstrapped genes to `nGenes =``r nGenes$ERVHg`

```{r bootERVHg, cache = TRUE}
bootERVHgOrtho <- mclapply(testVals, 
												bootMeans, 
												data = ERVHgData, 
												testIds = filter(ERVHgData, nonRTEERVHg)$ID, 
												refIds = filter(ERVHgData, !nonRTEERVHg)$ID,
												binCol = "elBins",
												nBoot = nBoot, 
												nGenes = nGenes$ERVHg,
												maxP = 0.01,
												mc.cores = nThreads)
names(bootERVHgOrtho) <- testVals
save(bootERVHgOrtho, file = "bootnonRTEERVHgOrtho.RData")
```

```{r, fig.cap = "**ERV Elements**. Visualisation of the differences between replicates within each tissue type. The top row indicates differences between samples for detection of expression, whilst the bottom row indicates differences in expression levels."}
bootERVHgOrtho %>%
	lapply(extract2, "samples") %>%
	as.data.frame() %>%
	melt() %>%
	as_data_frame() %>%
	mutate(type = gsub("hgs.(br|ht|lv|kd|ts).+", "\\1", variable),
				 TF = c("Expression", "Detection")[grepl("TF", variable) + 1],
				 sample = gsub("_TF", "", variable)) %>%
	filter(type != "logLength") %>%
	ggplot(aes(sample, value, fill = type)) +
	geom_hline(yintercept = 0) +
	geom_boxplot() +
	facet_grid(TF ~ type, scales = "free", space = "free_x", labeller = tissueLabels) +
	labs(title = "Ortholog: ERV Elements",
			 x = "Sample",
			 y = "Change") +
	guides(fill = FALSE) 
```

# Confidence Intervals for Length Differences

This step is a control to check that the length distributions were suitably paired.
Any intervals that don't contain zero will indicate that the two sets of genes will have different length distributions and the comparison will need to be reanalysed

```{r allBoot}
allBoot<- list( LINE = bootLINEHgOrtho, 
							 SINE = bootSINEHgOrtho, 
							 ERV = bootERVHgOrtho)
```


```{r confIntLen}
confIntLen <- allBoot %>%
  lapply(extract, "logLength") %>%
  lapply(getCI, alpha = 0.05, combined = TRUE) %>%
  bind_rows %>%
  mutate(TE = rep(elements$type))
```

```{r figS5, echo=FALSE, include=TRUE, fig.width=8}
figS5 <- confIntLen %>%
  mutate(TE = factor(TE, rev(unique(TE)))) %>%
  ggplot(aes(x = median, y = TE)) +
  geom_errorbarh(aes(xmin = lower, xmax = upper), height = 0.5) +
 # facet_wrap(~Region, nrow=1, scales = "free_x") +
  geom_vline(xintercept = 0, colour = "grey50", linetype = 2) +
  theme_bw() +
  labs(x = "95% CI for the diference in log(Length)") 
figS5
```

As type II errors are the primary concern here, no adjustment was applied for multiple comparisons.
Zero didn't lie outside of any of these intervals, and as such the subsequent results for the bootstrap can be deemed to be valid with gene legth not playing any role in the results.

```{r exportFigS5, echo = FALSE, include=TRUE, message=FALSE}
png("FigureS5.png", height = 10, width = 14.5, units = "cm", res = 300)
fs <- 9
figS5 +
theme(axis.text = element_text(size = fs - 0.5),
axis.title = element_text(size = fs + 1),
strip.text = element_text(size = fs + 1))
dev.off()
```


# Confidence Intervals for Elements & Regions

## Analysis by individual tissue

Here, all samples from a tissue type were merged for generation of the Bonferroni-adjusted 95% CIs.

```{r}
allCI <- allBoot %>%
  names() %>%
  lapply(function(x){
    lapply(allBoot[[x]], extract2, "samples") %>% 
      as_data_frame() %>%
      melt(variable.name = "Sample") %>%
      mutate(Class = c("Expression", "Detection")[grepl("TF$", Sample) + 1],
             Tissue = gsub(".+(br|ht|lv|kd|ov|ts).+", "\\1", Sample),
             Element = x) %>%
      as_data_frame() %>%
      filter(Tissue != "logLength") %>%
      group_by(Element, Tissue, Class) %>%
      summarise(lwr = quantile(value, probs = alpha / 2),
                med = median(value),
                upr = quantile(value, probs = 1 - alpha / 2),
                # The next two give the 95% CI without Bonferroni adjustment
                lwr95 = quantile(value, probs = 0.05 / 2), 
                upr95 = quantile(value, probs = 1 - 0.05 / 2))
  }) %>%
bind_rows() %>%
ungroup() %>%
mutate(Sig = sign(lwr) == sign(upr),
Tissue = unlist(tissueLabels(Tissue)),
Element = gsub("(recentTE)", "", Element))
```


### Differences in the Probability of Gene Expression

```{r, fig.cap = "Differences in the proability of a gene being considered as expressed, separated by tissue."}
allCI %>%
	filter(Class == "Detection") %>%
	ggplot(aes(med, Tissue, colour = Sig)) +
	geom_vline(xintercept = 0, colour = "blue", linetype = 2) +
	geom_point() +
	geom_errorbarh(aes(xmin = lwr, xmax = upr)) +
	scale_colour_manual(values = c("black", "red")) +
	facet_grid(~Element, scales = "free_x") +
	labs(x = "Difference in p(Expression > 0)") +
	guides(colour = FALSE)
```

### Differences in the Level of Gene Expression

```{r, fig.cap = "Differences in the proability of a gene being considered as expressed, separated by tissue."}
allCI %>%
  filter(Class == "Expression") %>%
  ggplot(aes(med, Tissue, colour = Sig)) +
  geom_vline(xintercept = 0, colour = "blue", linetype = 2) +
  ggtitle("Human") +
  geom_point() +
  geom_errorbarh(aes(xmin = lwr, xmax = upr)) +
  scale_colour_manual(values = c("black", "red")) +
  facet_grid(~Element, scales = "free_x") +
  labs(x = "Difference in Expression Levels") +
  guides(colour = FALSE)
```

## Analysis by combined tissues & samples

Here all samples and tissues were combined to create a general Bonferroni-adjusted 95% CI across all tissues.

```{r}
m <- nRegions * nElements
alpha <- 0.05 / m
mergedCI <- allBoot %>%
	names() %>%
	lapply(function(x){
		lapply(allBoot[[x]], extract2, "samples") %>% 
			as_data_frame() %>%
			melt(variable.name = "Sample") %>%
			mutate(Class = c("Expression", "Detection")[grepl("TF$", Sample) + 1],
						 Tissue = gsub(".+(br|ht|lv|kd|ts).+", "\\1", Sample),
						 Element = x) %>%
			as_data_frame() %>%
			filter(Tissue != "logLength") %>%
			group_by(Element, Class) %>%
			summarise(lwr = quantile(value, probs = alpha / 2),
								med = median(value),
								upr = quantile(value, probs = 1 - alpha / 2))
	}) %>%
	bind_rows() %>%
	ungroup() %>%
	mutate(Sig = sign(lwr) == sign(upr),
				 Region = gsub(".+(Utr3|Utr5|Prox|down)", "\\1", Element),
				 Region = gsub("Utr(3|5)", "\\1'UTR", Region),
				 Region = gsub("Prox", "Proximal Promoter", Region),
				 Region = gsub("down", "Downstream Region", Region),
				 Region = factor(Region, levels = c("Proximal Promoter", "5'UTR", "3'UTR", "Downstream Region")),
				 Element = gsub("(Utr3|Utr5|Prox|down)", "", Element))
```


### Differences in the Probability of Gene Expression

```{r, fig.cap = "Differences in the proability of a gene being considered as expressed, separated by tissue."}
mergedCI %>%
	filter(Class == "Detection") %>%
	ggplot(aes(med, Element)) +
	geom_vline(xintercept = 0, colour = "blue", linetype = 2) +
	geom_point() +
	geom_errorbarh(aes(xmin = lwr, xmax = upr)) +
	facet_wrap(~Region, scales = "free_x", nrow = 1) +
	labs(x = "Difference in p(Expression > 0)")
```

### Differences in the Level of Gene Expression

```{r, fig.cap = "Differences in the proability of a gene being considered as expressed, separated by tissue."}
mergedCI %>%
	filter(Class == "Expression") %>%
	ggplot(aes(med, Element, colour = Sig)) +
	geom_vline(xintercept = 0, colour = "blue", linetype = 2) +
	geom_point() +
	geom_errorbarh(aes(xmin = lwr, xmax = upr)) +
	scale_colour_manual(values = c("black", "red")) +
	facet_wrap(~Region, scales = "free_x", nrow = 1) +
	labs(x = "Difference in Expression Levels") +
	guides(colour = FALSE)
```

# Conversion to p-values

As the confidence intervals were regularly very close to excluding zero, it was coonsidered that the Bonferroni adjustment may be too stringent, and whilst reducing the number of Type I errors (i.e. false positives), this may be at the expense of an increase in Type II errors (i.e. false negatives).
As such, the level $\alpha$ at which the confidence interval excluded zero was found, and this was taken as representative of the p-value.
This would allow for adjustment using less restrictive strategies, such as Holm's method.

## Individual Tissues

```{r}
allP <- allBoot %>%
	names() %>%
	lapply(function(x){
		lapply(allBoot[[x]], extract2, "samples") %>% 
			as_data_frame() %>%
			melt(variable.name = "Sample") %>%
			mutate(Class = c("Expression", "Detection")[grepl("TF$", Sample) + 1],
						 Tissue = gsub(".+(br|ht|lv|kd|ts).+", "\\1", Sample),
						 Element = x) %>%
			as_data_frame() %>%
			filter(Tissue != "logLength") %>%
			arrange(value) %>%
			group_by(Element, Class, Tissue) %>%
			summarise(n = n(),
								p = (1 + findInterval(0, value)) / n) %>%
			ungroup() %>%
			rowwise() %>%
			mutate(p = 2*min(p, 1-p)) %>%
			ungroup()
	}) %>%
	bind_rows() %>%
	split(f = .$Class) %>%
	lapply(mutate, adjP = p.adjust(p, "holm"), fdr = p.adjust(p, "fdr"))
```

Unfortunately, Holm's method failed to increase the number of positive findings. 
However, the use of the FDR adjustment revealed a handful of further potential discoveries.

```{r}
allP$Detection %>%
	arrange(p) %>%
	extract(1:10,) %>%
	set_colnames(c("Element", "Type", "Tissue", "nSamples", "p", "$\\p_{adj}$", "FDR")) %>%
	pander(justify = "lllrrrr",
				 caption = "The 10 most highly ranked elements for changes in the probability of detection")
```

```{r}
allP$Expression %>%
	arrange(p) %>%
	extract(1:10,) %>%
	set_colnames(c("Element", "Type", "Tissue", "nSamples", "p", "$\\p_{adj}$", "FDR")) %>%
	pander(justify = "lllrrrr",
				 caption = "The 10 most highly ranked elements for changes in expression level")
```

## Combined Tissues

Combining bootstrap samples over tissues failed to yield any further discoveries

```{r}
allBoot %>%
	names() %>%
	lapply(function(x){
		lapply(allBoot[[x]], extract2, "samples") %>% 
			as_data_frame() %>%
			melt(variable.name = "Sample") %>%
			mutate(Class = c("Expression", "Detection")[grepl("TF$", Sample) + 1],
						 Tissue = gsub(".+(br|ht|lv|kd|ts).+", "\\1", Sample),
						 Element = x) %>%
			as_data_frame() %>%
			filter(Tissue != "logLength") %>%
			arrange(value) %>%
			group_by(Element, Class) %>%
			summarise(n = n(),
								p = (1 + findInterval(0, value)) / n) %>%
			ungroup() %>%
			rowwise() %>%
			mutate(p = 2*min(p, 1-p)) %>%
			ungroup()
	}) %>%
	bind_rows() %>%
	split(f = .$Class) %>%
	lapply(mutate, adjP = p.adjust(p, "holm"), fdr = p.adjust(p, "fdr")) %>%
	lapply(arrange, p) %>%
	lapply(pander)
```


# Appendix

```{r}
save.image("SteveBootstrapOrtho-nsTE.RData")
```


```{r}
sessionInfo()
```

